src/couch_expiring_cache/src/couch_expiring_cache_fdb.erl

```diff
@@ -0,0 +1,107 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(couch_expiring_cache_fdb).
+
+-export([
+    insert/5,
+    lookup/2,
+    clear_expired_range/2
+]).
+
+
+-define(XC, 53). % coordinate with fabric2.hrl
+-define(PK, 1).
+-define(EXP, 2).
+
+
+-include_lib("couch_expiring_cache/include/couch_expiring_cache.hrl").
+
+
+% Data model
+% see: https://forums.foundationdb.org/t/designing-key-value-expiration-in-fdb/156
+%
+% (?XC, ?PK, Name, Key) := (Val, StaleTS, ExpireTS)
+% (?XC, ?EXP, ExpireTS, Name, Key) := ()
+
+
+-spec insert(Name :: binary(), Key :: binary(), Value :: binary(),
+    StaleTS :: millisecond(), ExpiresTS :: millisecond()) -> ok.
+insert(Name, Key, Val, StaleTS, ExpiresTS) ->
+    LayerPrefix = couch_expiring_cache_server:layer_prefix(),
+    PK = primary_key(Name, Key, LayerPrefix),
+    PV = erlfdb_tuple:pack({Val, StaleTS, ExpiresTS}),
+    XK = expiry_key(ExpiresTS, Name, Key, LayerPrefix),
+    XV = erlfdb_tuple:pack({}),
+    fabric2_fdb:transactional(fun(Tx) ->
+        ok = erlfdb:set(Tx, PK, PV),
+        ok = erlfdb:set(Tx, XK, XV)
+    end).
+
+
+-spec lookup(Name :: binary(), Key :: binary()) ->
+    not_found | {fresh, Val :: binary()} | {stale, Val :: binary()} | expired.
+lookup(Name, Key) ->
+    LayerPrefix = couch_expiring_cache_server:layer_prefix(),
+    PK = primary_key(Name, Key, LayerPrefix),
+    fabric2_fdb:transactional(fun(Tx) ->
+        case erlfdb:wait(erlfdb:get(Tx, PK)) of
+            not_found ->
+                not_found;
+            Bin when is_binary(Bin) ->
+                {Val, StaleTS, ExpiresTS} = erlfdb_tuple:unpack(Bin),
+                Now = erlang:system_time(?TIME_UNIT),
+                if
+                    Now < StaleTS -> {fresh, Val};
+                    Now < ExpiresTS -> {stale, Val};
+                    true -> expired
+                end
+        end
+    end).
+
+
+-spec clear_expired_range(EndTS :: millisecond(), Limit :: non_neg_integer()) ->
+    OldestTS :: integer().
+clear_expired_range(EndTS, Limit) when Limit > 0 ->
+    LayerPrefix = couch_expiring_cache_server:layer_prefix(),
+    ExpiresPrefix = erlfdb_tuple:pack({?XC, ?EXP}, LayerPrefix),
+    Callback = fun
```

Comment:
Maybe not worth having a callback to compute the oldest Ts. A `case {ExpiresTs, Acc} of ...` might not be too bad.

Reply:
I tried a case statement, but it seemed pretty verbose. I ended up replacing the fun with a function named `oldest_ts/2` which seems to both show intent and be most compact.